api/src/main/java/io/grpc/Detachable.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * A <i>Detachable</i> encapsulates an object that can be forked with underlying resources
+ * detached and transferred to a new instance. The forked instance takes over the ownership
+ * of resources and is responsible for releasing after use. The forked instance preserves
+ * states of detached resources. Resources can be consumed through the forked instance as if
+ * being continually consumed through the original instance. The original instance discards
+ * states of detached resources and is no longer consumable as if the resources are exhausted.
+ */
+@ExperimentalApi("TODO")
+public interface Detachable {
+
+  /**
+   * Fork a new instance with underlying resources detached from this instance and
+   * transferred to the new instance.
+   *
+   * @throws IllegalStateException if the underlying resources have already been detached.
+   */
+  public Detachable detach();
```

Comment:
Hmm... I see how the typing is awkward. It is probably more useful for us to return InputStream here.

Reply:
> Returning InputStream is definitely great for typing. But doesn't it limit this API to InputStream only? Is that desirable for us/future? Would need to update the overall doc/specification if this only applies to InputStream.

Yes, maybe, yes :smile:. Yeah, it'd be nice if generics could manage this, but since this is a tagging interface the caller will be casting which makes generics useless.

I'm not all that concerned about making it generic and usable with any types; I'm fine with focusing on InputStream. My comment about returning InputStream, is simply that Detachable by itself is a useless type to return; you can't do anything with it. That said, maybe you are right that it conveys "you get the same type of thing" a bit more. In either case, we might want to focus the documentation on InputStream more so that it is more clear what its purpose is and how it is used.

Reply:
Updated.